torchaudio/torch_overdrive.cpp

```diff
@@ -0,0 +1,52 @@
+#include <torch/extension.h>
+
+namespace torch {
+namespace audio {
+
+template <typename scalar_t>
+void overdrive_cpu_kernel(
+    at::TensorAccessor<scalar_t, 2> waveform_accessor,
+    at::TensorAccessor<scalar_t, 2> temp_accessor,
+    at::TensorAccessor<scalar_t, 1> last_in_accessor,
+    at::TensorAccessor<scalar_t, 1> last_out_accessor,
+    at::TensorAccessor<scalar_t, 2> output_waveform_accessor) {
+  int64_t n_frames = waveform_accessor.size(1);
+  int64_t n_channels = waveform_accessor.size(0);
+
+  for (int64_t i_channel = 0; i_channel < n_channels; ++i_channel) {
```

Comment:
Depending on the amount of work you might benefit from using [parallel_for](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Parallel.h#L53).

Most PyTorch CPU operators are parallelized, unless there's no obvious need due to memory-boundedness.

Another issue with pure C for C++ extensions, for now, is autovectorization. We can't ship avx2 code without a CPU capability based dispatch. That means for C code in extensions like this we're for now restricted to SSE and related.

Of course this is taken care of when you call into at:: operations directly, since they each take advantage being part of libtorch.

Reply:
@cpuhrsch , yeah parallelization can be applied only for the channels loop .  I was not sure how the parallel_for treats the inner sequential  loop , so kept it without the parallel_for. A parallel thread won't interfere with other parallel thread's inner loop working right ?

Reply:
@bhargavkathivarapu - I'm not sure what you mean by "interfere with" exactly. As in, shared variables or creating integers etc.?